File: BOJ/1874.cpp
// 1874번_스택 수열
#include <iostream>
#include <stack>
#include <vector>
using namespace std;
int main(){
    int n, idx = 0, seq[100001] = {0, };
    stack<int> st;
    vector<char> v;
    cin >> n;
    for(int i=0; i<n; ++i) cin >> seq[i];
    // 1~n까지 스택에 push하며 pop조건을 만족하는지 확인
    // 스택 top이 입력된 수열의 idx번째 수와 같으면 pop
    for(int i=1; i<=n; ++i){
        st.push(i);
        v.push_back('+');     
        // 연속 pop도 가능하므로 반복문으로 처리 
        while(!st.empty() && (st.top() == seq[idx])){
            st.pop();
            v.push_back('-');
            idx++;
        }
    }
    if(!st.empty()) cout << "NO";
    else{
        for(int i=0; i<v.size(); ++i) 
            cout << v[i] << "\n";
    }    
    return 0;
}

File: BOJ/1874.py
# 1874번_스택 수열
import sys
input = sys.stdin.readline
n = int(input())
seq = [0] * n
for i in range(n):
    seq[i] = int(input())

# 자연수 차례대로 삽입하다가 현재 순서에 맞는 수 찾으면 삭제
ans = []
st = []
idx = 0 # seq 배열의 인덱스
for i in range(1, n+1):
    st.append(i)
    ans.append('+')

    # 스택 top에서 원하는 수 만나는동안 계속 pop
    while st and (st[-1]>=seq[idx]):
        st.pop()
        ans.append('-')
        idx += 1

# 스택이 비어있으면 입력된 수열을 만들수 있다는 의미
if not st:
    for op in ans: print(op)
else:
    print('NO')